bothan-api/server-cli/src/commands/query.rs

```diff
@@ -0,0 +1,466 @@
+    opts: bothan_coingecko::WorkerOpts,
+    query_ids: &[String],
+    timeout_interval: T,
+) -> anyhow::Result<()> {
+    let api = bothan_coingecko::api::RestApiBuilder::new(opts.url, opts.user_agent, opts.api_key)
+        .build()?;
+
+    let distinct_ids = get_distinct_ids(query_ids);
+    let asset_infos = timeout(
+        timeout_interval.into(),
+        RestAssetInfoProvider::get_asset_info(&api, &distinct_ids),
+    )
+    .await??;
+
+    print_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_coinmarketcap<T: Into<Duration>>(
+    opts: bothan_coinmarketcap::WorkerOpts,
+    query_ids: &[String],
+    timeout_interval: T,
+) -> anyhow::Result<()> {
+    let api = bothan_coinmarketcap::api::RestApiBuilder::new(opts.url, opts.api_key).build()?;
+
+    let distinct_ids = get_distinct_ids(query_ids);
+    let asset_infos = timeout(
+        timeout_interval.into(),
+        RestAssetInfoProvider::get_asset_info(&api, &distinct_ids),
+    )
+    .await??;
+
+    print_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_htx<T: Into<Duration>>(
+    opts: bothan_htx::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = bothan_htx::api::WebSocketConnector::new(opts.url);
+    let mut provider = connector.connect().await?;
+
+    let distinct_ids = get_distinct_ids(query_ids);
+    WebSocketAssetInfoProvider::subscribe(&mut provider, &distinct_ids).await?;
+    let asset_infos = query(&mut provider, distinct_ids, timeout.into()).await?;
+
+    print_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_kraken<T: Into<Duration>>(
+    opts: bothan_kraken::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = bothan_kraken::api::WebSocketConnector::new(opts.url);
+    let mut provider = connector.connect().await?;
+
+    let distinct_ids = get_distinct_ids(query_ids);
+    WebSocketAssetInfoProvider::subscribe(&mut provider, &distinct_ids).await?;
+    let asset_infos = query(&mut provider, distinct_ids, timeout.into()).await?;
+
+    print_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_okx<T: Into<Duration>>(
+    opts: bothan_okx::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = bothan_okx::api::WebSocketConnector::new(opts.url);
+    let mut provider = connector.connect().await?;
+
+    let distinct_ids = get_distinct_ids(query_ids);
+    WebSocketAssetInfoProvider::subscribe(&mut provider, &distinct_ids).await?;
+    let asset_infos = query(&mut provider, distinct_ids, timeout.into()).await?;
+
+    print_asset_infos(asset_infos);
+    Ok(())
+}
+
+fn get_distinct_ids(ids: &[String]) -> Vec<String> {
+    let mut seen = HashSet::new();
+    let mut result = Vec::with_capacity(ids.len());
+
+    for id in ids {
+        if seen.insert(id) {
+            result.push(id.clone());
+        }
+    }
+
+    result
+}
```

Comment:
why do you need this? just use id.iter().dedup() if you need a deduplicated copy